### ISSUE TITLE
no-unused-prop-types false positives for HoCs (ex: react-redux connect)

### ISSUE THREAD

**Issue by rsolomon** [yannickcr/eslint-plugin-react#1448]
If you have a set of props that are only used in functions passed to an HoC, `no-unused-prop-types` will erroneously say that the props are unused. This is a major issue for any app that uses react-redux.

Example component that fails:

```
type User = { name: string };
type AppState = { [id: string]: Array<User> };
type OwnProps = { id: string };
type PropsFromState = { users: Array<User> };

function mapStateToProps(state: AppState, ownProps: OwnProps): PropsFromState {
  const { id } = ownProps;
  return state[id];
}

class UsersViewer extends React.Component<OwnProps & PropsFromState> {
  render(): React.Node {
    const { users } = this.props;
    return <ul>{users.map(user => <li>user.name</li>)}</ul>;
  }
}

export default connect(mapStateToProps)(UsersViewer);
```

Even though the `id` prop is used in `mapStateToProps`, the following error will still throw:

```
error  'id' PropType is defined but prop is never used  react/no-unused-prop-types
```

This is an issue in v7.4.0 only

**Comment by ljharb** [comment]
It's an issue always, it's just that now we detect prop types better.

There's really no way around this without encoding special knowledge of `connect` in this plugin. I'd recommend using override comments for these cases.

**Comment by ardok** [comment]
@ljharb When you said override comments, do you mean the `/* eslint-disable react/no-unused-prop-types */` ?

**Comment by ljharb** [comment]
@ardok yes, exactly right - `// eslint-disable-line` or `// eslint-disable-next-line` are the more targeted ones.

**Comment by ardok** [comment]
Also, is destructuring from function param a known false positive as well?
```
onFilterChange = ({id}: {id: string}) => {
    this.setState({
      selectedId: id,
    });
};
```

That will complain about:
`error  'id' PropType is defined but prop is never used  react/no-unused-prop-types`

It's not even a "PropType".
I have the skip shape props on already: `"react/no-unused-prop-types": [2, { "skipShapeProps": true }],`

**Comment by ljharb** [comment]
That last one seems like a bug; can you file that separately with the full component code?

**Comment by ardok** [comment]
https://github.com/yannickcr/eslint-plugin-react/issues/1561

**Comment by anchaljain123** [comment]
I am having the trouble when I use props in helper function in a component and specified props validation for those props. Here is my code 
const specs = {
	drop(props, monitor, component) {
		document.getElementById(monitor.getItem().id).style.display = 'block';
		const { placeholderIndex } = component.state;
		const lastX = monitor.getItem().x;
		const lastY = monitor.getItem().y;
		const nextX = props.x;
		let nextY = placeholderIndex;
		if (props.cards.length >= props.limit && lastX !== nextX) {
			return;
		}
		if (lastY > nextY) { // move top
			nextY += 1;
		} else if (lastX !== nextX) { // insert into another list
			nextY += 1;
		}

		if (lastX === nextX && lastY === nextY) { // if position equel
			return;
		}
		props.moveCard(lastX, lastY, nextX, nextY);
	},
	hover(props, monitor, component) {
		// defines where placeholder is rendered
		const placeholderIndex = getPlaceholderIndex(
			monitor.getClientOffset().y,
			findDOMNode(component).scrollTop
		);

		// horizontal scroll
		if (!props.isScrolling) {
			if (window.innerWidth - monitor.getClientOffset().x < 200) {
				props.startScrolling('toRight');
			} else if (monitor.getClientOffset().x < 200) {
				props.startScrolling('toLeft');
			}
		} else if (window.innerWidth - monitor.getClientOffset().x > 200 &&
			monitor.getClientOffset().x > 200
		) {
			props.stopScrolling();
		}
	}
};
export default class Cards extends Component {
....
}
Cards.propTypes = {
	connectDropTarget: PropTypes.func.isRequired,
	moveCard: PropTypes.func.isRequired,
	cards: PropTypes.array.isRequired,
	x: PropTypes.number.isRequired,
	isOver: PropTypes.bool,
	item: PropTypes.object,
	canDrop: PropTypes.bool,
	startScrolling: PropTypes.func,
	stopScrolling: PropTypes.func,
	isScrolling: PropTypes.bool,
	name: PropTypes.string,
	limit: PropTypes.number,
	removeCard: PropTypes.func,
	itemStyle: PropTypes.object
};
It still throws me es lint error of like 'error  'moveCard' PropType is defined but prop is never used        react/no-unused-prop-types'

**Comment by ljharb** [comment]
@anchaljain123 passing around the full `props` object is an antipattern, and also makes it unlintable. Try destructuring all the props you need off of it first, before you pass it into any of the `specs` functions.

**Comment by anchaljain123** [comment]
My 'specs' are parameter to dragSource like 
@DropTarget('card', specs, (connectDragSource, monitor) => ({
	connectDropTarget: connectDragSource.dropTarget(),
	isOver: monitor.isOver(),
	canDrop: monitor.canDrop(),
	item: monitor.getItem()
}))
so I cannot pass anything from 'specs'
